### PR TITLE
Add keycloak authorization server and setup spring security

### LINKED ISSUE #6: Implement security layer
- [x] use oauth2 protocol
- [x] use one of oauth servers keycloak / auth0 / spring 
- [x] protect endpoints with token authorization 
- [x] add 3rd party login providers  e.g. github

### PR DESCRIPTION
**Pull request by brutovsky**
Protect todo-service with oauth2 authentication.
Update tests.

### CODE REVIEW

**Comment by brutovsky** [comment]
Closes #6

File: docker-compose.yml [hunk 1]
```diff
@@ -20,5 +21,45 @@ services:
     depends_on:
       - todo-service-database
 
+  todo-service-api-2:
```
Comment (guga4ka) [inline docker-compose.yml]: Remove duplication of service. Use `docker compose --sacle` instead

File: keycloak/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,6 @@
+FROM jboss/keycloak:16.0.0
+
+COPY ./micro2do-realm.json /tmp/realm-export.json
```
Comment (guga4ka) [inline keycloak/Dockerfile]: it better to use docker compose and mound these files and don't spend time building a docker image

File: todo-service/src/main/java/com/brtvsk/todoservice/controller/TodoController.java [hunk 3]
```diff
@@ -38,11 +39,11 @@ public TodoController(final TodoService todoService) {
     }
     )
     @GetMapping
-    public List<TodoResponse> all(Optional<Boolean> done) {
+    public List<TodoResponse> all(Optional<Boolean> done, final Authentication authentication) {
```
Comment (guga4ka) [inline todo-service/src/main/java/com/brtvsk/todoservice/controller/TodoController.java]: Replace `Authentication` with custom user model

**Review by guga4ka — DISMISSED** [review]

File: docker-compose.yml (hunk 1 above)
Comment (brutovsky) [reply docker-compose.yml]: Removed duplicates

File: keycloak/Dockerfile (hunk 2 above)
Comment (brutovsky) [reply keycloak/Dockerfile]: Moved keycloack configuration to docker-compose

File: todo-service/src/main/java/com/brtvsk/todoservice/controller/TodoController.java (hunk 3 above)
Comment (brutovsky) [reply todo-service/src/main/java/com/brtvsk/todoservice/controller/TodoController.java]: Added AuthUser model

**Review by guga4ka — APPROVED** [review]
lgtm

**Pull request merged**